agent/spec/lib/kontena/observable_spec.rb

```diff
@@ -1,166 +1,275 @@
+      observer_actor.ping
+
+      expect(subject).to be_observed
+
+      expect{observer_actor.crash}.to raise_error(RuntimeError)
+
+      # make sure the observer is really dead
+      expect{observer_actor.ping}.to raise_error(Celluloid::DeadActorError)
+      expect(observer_actor).to_not be_alive
 
-    observers.each do |obs|
-      obs.async.start
+      subject.update(value)
+      expect(subject).to_not be_observed
     end
 
-    # run updates sync while the observers are starting
-    subject.spam_updates(1..update_count, interval: 0.001)
+    it "delivers updates in the right order" do
+      observer_actor.async.test_ordering(subject)
 
-    # wait for observable to notify all observers
-    subject.ping
+      update_count = 150
 
-    # wait for all observers to observe and update
-    observers.each do |obs| obs.ping end
-    observers.each do |obs| obs.ping end # and maybe a second round for the async update
+      observable_actor.spam_updates(1..update_count, interval: false)
 
-    # all observers got the final value
-    expect(observers.map{|obs| obs.value}).to eq [update_count] * observer_count
+      expect(observer_actor.observed_values.last).to eq update_count
+      expect(observer_actor.ordered?).to be_truthy
+    end
 
-    # this is potentially racy, but it's important, or this spec doesn't test what it should
-    expect(observers.map{|obs| obs.first}.max).to be > 1
+    it "handles concurrent observers" do
+      observer_count = 20
+      update_count = 10
 
-    # also expect this...
-    expect(observers.map{|obs| obs.ordered?}).to eq [true] * observer_count
-  end
+      # setup
+      observer_actors = observer_count.times.map {
+        TestObserverActor.new
+      }
+
+      # start spamming updates while the observer actors are concurrently observing
+      future = observable_actor.future.spam_updates(1..update_count, delay: 0.005, interval: 0.001)
+
+      observer_actors.each do |actor|
+        actor.async.test_ordering(subject)
+        sleep 0.001
+      end
+
+      # wait for observable actor to notify all observers
+      future.value
+
+      # wait for all observers to observe and update
+      observer_actors.each do |actor|
+        actor.ping
+      end
+
+      # all observers get the final value
+      expect(observer_actors.map{|actor| actor.observed_values.last}).to eq [update_count] * observer_count
+
+      # make sure they get the values in order
+      expect(observer_actors.map{|actor| actor.ordered?}).to eq [true] * observer_count
+
+      # some observers observed before the first update, some after
+      # this is potentially racy, but it's important, or this spec doesn't test what it should!
+      min_first = observer_actors.map{|actor| actor.observed_values.first}.min
+      max_first = observer_actors.map{|actor| actor.observed_values.first}.max
+
+      expect(min_first).to be < max_first
+      expect(max_first).to be > 1
```

Comment:
These spec failures are not actually observable bugs: they just mean that the spec conditions did not trigger the expected race conditions. It should probably retry instead of failing the spec: if there's a 1% chance of missing the race, then retrying three times should be a 0.001% chance.

```
  1) Kontena::Observable registered by an Actor handles concurrent observers

     Failure/Error: expect(min_first).to be < max_first

       expected: < 1

            got:   1

     # ./spec/lib/kontena/observable_spec.rb:226:in `block (3 levels) in <top (required)>'
```

Reply:
This seems to occasionally fail in travis, so I think it needs the retry loop to not become a flaky spec.